### PR TITLE
perf(storage): Merge multiple imms in the staging version to a large one

### PR DESCRIPTION
**Pull request by StrikeW**
I hereby agree to the terms of the [Singularity Data, Inc. Contributor License Agreement](https://gist.github.com/skyzh/0663682a70b0edde7ae991492f2314cb#file-s9y_cla).

## What's changed and what's your intention?
The problem is described in https://github.com/risingwavelabs/risingwave/issues/5634.
We want to merge the staging imms of a state table to reduce the number of iterators needed on the merge read procedure. So we extend the structure of `SharedBufferBatch` (the Imm) to accommodate data from multiple epochs. 

Hummock uploader will group ingested batches by their table shard (`imms_by_table_shard`) and is responsible to start merging tasks on each `SealEpoch` event. Uploader will manage the lifecycle of merging tasks. For example, it will poll those tasks and drop them when a `flush` is requested. Because the **merge-imms** optimization only can improve in-memory scenarios, if the `flush` is requested by `spill` the system may be short of memory. And dropping merging task when sync occurred is to keep simplicity. In addition, we only merge sealed data which means a merging task would not touch unsealed data and is always in the scope of a checkpoint.

When a merging task is finished, the uploader will emit an `ImmMerged` event, the event handler will update the staging version of the corresponding state table and remove Imms that have been merged from the `sealed_data`. The merged Imm will be fed into the `flush_imms` procedure for compaction.

Bench result: https://www.notion.so/risingwave-labs/Bench-Results-for-Merge-Imms-256a1fa4937a4a0d91d7de9e264ba32c

## Checklist

- [x] I have written necessary rustdoc comments
- [x] I have added necessary unit tests and integration tests
~- [ ] I have added fuzzing tests or opened an issue to track them. (Optional, recommended for new SQL features).~
- [x] All checks passed in `./risedev check` (or alias, `./risedev c`)

## Refer to a related PR or issue link (optional)
https://github.com/risingwavelabs/risingwave/issues/5634

### CODE REVIEW

File: src/storage/src/hummock/store/state_store.rs
Comment (Little-Wallace) [inline src/storage/src/hummock/store/state_store.rs:104]: Is it repeated with `ImmMergeTask::new`

**Comment by Little-Wallace** [comment]
It is no need to send `SharedBufferBatch` to `event handler` and then put them into tokio pool. We can merge them in the streaming thread directly. It would be simple and effective.

```rust
fn ingest_batch(&self, batch: SharedBufferBatch) {
     let imms = self.read_version.read().get_imms();
     let mut build_imms = vec![]
     for imm in imms.iter().rev() {
          // do some strategy check
     }
     let new_mem = ImmutableMemtable::build(build_imms);
     self.read_version.write().set_imms(new_mem);
}
```

**Comment by StrikeW** [comment]
> It is no need to send `SharedBufferBatch` to `event handler` and then put them into tokio pool. We can merge them in the streaming thread directly. It would be simple and effective.
> 
> 
> 
> ```rust
> 
> fn ingest_batch(&self, batch: SharedBufferBatch) {
> 
>      let imms = self.read_version.read().get_imms();
> 
>      let mut build_imms = vec![]
> 
>      for imm in imms.iter().rev() {
> 
>           // do some strategy check
> 
>      }
> 
>      let new_mem = ImmutableMemtable::build(build_imms);
> 
>      self.read_version.write().set_imms(new_mem);
> 
> }
> 
> ```

Yes. I also prefer to evaluate sync merge first. But there is still a bug I am tracking.

**Comment by StrikeW** [comment]
Hi, this PR has two unsolved problems. 1) We should not block the hummock event loop. 2) It is better to drop imms that have been merged to reduce memory footprint and feed the merged_imm to `flush_imms` procedure.

In the current implementation, the hummock uploader will build a future to merge imms for each local hummock instance and await the future in the hummock event loop, which will block the event loop. But if we don't await the future immediately, then we cannot make sure the merge is finished before the flush (especially the flush triggered by **checkpoint**), which will introduce two challenges:
1. We need to cancel unfinished merge tasks triggered before the checkpoint, which will need additional effort to maintain consistency
2. We need to feed imm and merged_imm as input to `flush_imms`, which will make the flush procedure become complex

To simplify the implementation, I want to guarantee the merge of multiple imms is finished **before** flush/checkpoint. A feasible solution is to do the merge in the executor ingest_batch procedure instead of the hummock uploader. What do you think?

**Comment by wenym1** [comment]
> 1. We need to cancel unfinished merge tasks triggered before the checkpoint, which will need additional effort to maintain consistency.

If the merge task can be designed to be with no side effect, and then we can abort the merge task without fear.

> 2. We need to feed imm and merged_imm as input to flush_imms, which will make the flush procedure become complex.

If the difference between imm and merged_imm is only that the merged_imm holds multiple epochs while imm holds only one epoch, we can unify them to a same struct and we don't have to distinguish between imm and merged_imm.

If the two mentioned challenges can be solved, I think we can still try to merge the imm asynchronously without affecting the  `ingest_batch`.

I think the flush procedure can be not that complex. We only need to maintain one more state in the uploader, so call `MergingTask`. And when we poll the next event of the uploader, we can also poll the join handle of the merge task and emit event `ImmMerged` when a merged task is finished. For flushing, if we want to start a flush task that depends on a merging task, we can abort the merging task,  and use the original input of the merging task as the input of the flush task. For sync, if the sync task depends on the output of some merge tasks, we can put the logic of awaiting for the join handle of merging task in the spawned future of the sync task, and then the uploader will not have to handle the logic of polling these merge tasks, and only need to poll the spawned future of the sync task.

**Comment by StrikeW** [comment]
Thanks for your suggestions. I will go for this direction and try to make it.

File: src/storage/src/hummock/shared_buffer/shared_buffer_batch.rs
Comment (hzxa21) [inline src/storage/src/hummock/shared_buffer/shared_buffer_batch.rs:960]: I don't think we need a `SharedBufferBatchSnapshotIterator` for read epoch filtering since filtering invisible epoch is the responsibility of user key iterator. Although it is possible do epoch filtering push-down for all hummock iterators, we don't see the needs so we didn't do it (e.g. no epoch filter for SstableIterator, MergeIterator).

File: src/storage/src/hummock/shared_buffer/shared_buffer_batch.rs
Comment (hzxa21) [inline src/storage/src/hummock/shared_buffer/shared_buffer_batch.rs:368]: can we call this `min_epoch` instead?

File: src/storage/src/hummock/shared_buffer/shared_buffer_batch.rs
Comment (hzxa21) [inline src/storage/src/hummock/shared_buffer/shared_buffer_batch.rs:142]: shouldn't `imm_ids.iter().max().unwrap()` be sufficient since `imm_ids` cannot be empty?

File: src/storage/src/hummock/store/version.rs
Comment (hzxa21) [inline src/storage/src/hummock/store/version.rs:385]: This assertion seems unneccessary since the all condition is exactly the same as the retain condition above.

File: src/storage/src/hummock/store/version.rs
Comment (hzxa21) [inline src/storage/src/hummock/store/version.rs:373]: Is it possible for a mergfed_imm to spand across max_committed_epoch? For exmaple, mce=4 and merged_imm has epoch [3,4,5].

File: src/storage/src/hummock/store/version.rs
Comment (hzxa21) [inline src/storage/src/hummock/store/version.rs:550]: L546 filters imm whose min epoch <= `min_epoch`, which seems incorrect. For example, if we have a merged imm with `epochs = [1, 2, 3]` and the provided `min_epoch = 2`, we should keep this merged imm.

File: src/storage/hummock_test/src/hummock_read_version_tests.rs
Comment (StrikeW) [reply src/storage/hummock_test/src/hummock_read_version_tests.rs]: Removed. it is for imm merge in the `ingest_batch` path.

File: src/storage/src/hummock/store/version.rs
Comment (StrikeW) [reply src/storage/src/hummock/store/version.rs]: Refactored to configurable in `StorageConfig`.
Yes. The current strategy to trigger imm merge is straightforward, and I have observed that data size (operator states) of each epoch can be different. We can optimize the strategy further. But pick up a reasonable size threshold is also not trivial.

Comment (StrikeW) [reply src/storage/src/hummock/store/version.rs]: > In the current implementation, a merged IMM canont be merged again so we can still have performance issue if we have many small merged IMMs.

Hmm, I think a merged IMM could merge again. We do the merge for each table shard and obey the asc order of epochs, we can still append a merged_imm to the `imms_by_table_shard` container, then it can be merged with imms generated by newer epochs. Am I miss something?

File: src/storage/src/hummock/event_handler/uploader.rs
Comment (StrikeW) [reply src/storage/src/hummock/event_handler/uploader.rs:314]: I am working on unifying the `imms` and `imms_by_table_shard`.

File: src/storage/src/hummock/store/version.rs
Comment (StrikeW) [reply src/storage/src/hummock/store/version.rs]: It should not happen. Because we only merge sealed epoch data. So a merged_imm would not cross the checkpoint epoch boundary.

**Comment by StrikeW** [comment]
Hi, previous comments have been addressed. Would you like to take a second round of look? @hzxa21 @wenym1

File: src/storage/src/hummock/event_handler/uploader.rs
Comment (wenym1) [inline src/storage/src/hummock/event_handler/uploader.rs:369]: We may remove this comment.

File: src/storage/src/hummock/event_handler/uploader.rs
Comment (wenym1) [inline src/storage/src/hummock/event_handler/uploader.rs:456]: I think there is no any IO operation in the merge task. So the merge task should not get any error and should always return with success?

File: src/storage/src/hummock/event_handler/hummock_event_handler.rs
Comment (wenym1) [inline src/storage/src/hummock/event_handler/hummock_event_handler.rs:148]: We may not need the clone?

File: src/storage/src/hummock/shared_buffer/shared_buffer_batch.rs
Comment (wenym1) [inline src/storage/src/hummock/shared_buffer/shared_buffer_batch.rs:531]: As mentioned above, we may move the following logic to shared buffer compact.

File: src/storage/src/hummock/shared_buffer/shared_buffer_batch.rs
Comment (wenym1) [inline src/storage/src/hummock/shared_buffer/shared_buffer_batch.rs:637]: Any reason to adopt two separated idx for key and value? I think they are in the same progress even though we contains multiple epochs in the imm.

**Review by wenym1 — COMMENTED** [review]
I will continue to review more details of this PR.

File: src/storage/src/hummock/event_handler/uploader.rs
Comment (StrikeW) [reply src/storage/src/hummock/event_handler/uploader.rs:369]: fixed

File: src/storage/src/hummock/shared_buffer/shared_buffer_batch.rs
Comment (StrikeW) [reply src/storage/src/hummock/shared_buffer/shared_buffer_batch.rs:553]: After unifying the struct of merged_imm and imm, `SharedBufferBatch` becomes a two dimension array. The value index is for iterating different versions of a key on each `next()` call.

File: src/storage/src/hummock/event_handler/hummock_event_handler.rs
Comment (StrikeW) [reply src/storage/src/hummock/event_handler/hummock_event_handler.rs:453]: Don't understand the problem here. I already modified the `poll` of `NextUploaderEvent` to emit the `ImmMerged` event.

File: src/storage/src/hummock/event_handler/uploader.rs
Comment (StrikeW) [reply src/storage/src/hummock/event_handler/uploader.rs:130]: I added a ut `test_drop_success_merging_task` which can safely drop a successfully polled join handle.

File: src/storage/src/hummock/event_handler/uploader.rs
Comment (StrikeW) [reply src/storage/src/hummock/event_handler/uploader.rs:455]: Yes. Currently, the merging task would only return `Ok`, but I think it is not harmful to have a match arm for `Err` in this function.

File: src/storage/src/hummock/event_handler/hummock_event_handler.rs
Comment (wenym1) [reply src/storage/src/hummock/event_handler/hummock_event_handler.rs:453]: Currently when we handle the `ImmMerged` event, we call `self.uploader.update_sealed_data(...)`. However, the `ImmMerged` is coming out from the uploader. Therefore, we should not handle the `self.uploader.update_sealed_data(...)` in the event handler. Instead, we may handle `self.uploader.update_sealed_data(...)` before emitting the event, because the logic is internal for the uploader.

File: src/storage/src/hummock/event_handler/hummock_event_handler.rs
Comment (wenym1) [reply src/storage/src/hummock/event_handler/hummock_event_handler.rs:510]: I think a better solution is to handle the potential flush inside `start_merge_imms`, because if we really needs to spill the data, we will abort the merging task that just gets started, and if so we'd better not start the merging task at all.

File: src/storage/src/hummock/event_handler/hummock_event_handler.rs
Comment (StrikeW) [reply src/storage/src/hummock/event_handler/hummock_event_handler.rs:453]: Got it. fixed.

File: src/storage/src/hummock/event_handler/hummock_event_handler.rs
Comment (StrikeW) [reply src/storage/src/hummock/event_handler/hummock_event_handler.rs:507]: Good point. fixed. Now will acquire memory before generating a merging task. If acquire memory fail, it skips generate the task.

**Review by StrikeW — COMMENTED** [review]
Comments fixed. PTAL again, thank you. @wenym1

File: src/storage/src/hummock/compactor/shared_buffer_compact.rs
Comment (wenym1) [inline src/storage/src/hummock/compactor/shared_buffer_compact.rs:283]: We shall use `OrderedMergeIteratorInner` in case that we may support write-anytime in the future.

File: src/storage/src/hummock/compactor/shared_buffer_compact.rs
Comment (wenym1) [inline src/storage/src/hummock/compactor/shared_buffer_compact.rs:292]: Can we possibly avoid a memory allocation and copy here since what is stored in Imm is `Bytes`, which supports shallow copy on the byte payload.

Since the generic type of the merge iterator is unknown in the function scope, we don't have to depends on the method defined in the `HummockIterator` trait and instead we can implement some specific method for the merge iterator and the imm iterator.

File: src/storage/src/hummock/event_handler/hummock_event_handler.rs
Comment (wenym1) [inline src/storage/src/hummock/event_handler/hummock_event_handler.rs:456]: Shall we panic or add a log if the instance id is not found?

File: src/storage/src/hummock/shared_buffer/shared_buffer_batch.rs
Comment (wenym1) [reply src/storage/src/hummock/shared_buffer/shared_buffer_batch.rs:553]: The naming `value_idx` and `key_idx` is a little confused with the key and value for the iterator. Shall we have a less confusing naming?

File: src/storage/src/hummock/event_handler/uploader.rs
Comment (wenym1) [inline src/storage/src/hummock/event_handler/uploader.rs:393]: Shall we introduce an assumption that a merging task is only for a specific table id and instance id? If so we can avoid repeatedly calling `task.input_imms.entry(...)`

File: src/storage/src/hummock/event_handler/uploader.rs
Comment (wenym1) [inline src/storage/src/hummock/event_handler/uploader.rs:711]: Shall we have a log when we fail to require the memory?

**Review by wenym1 — COMMENTED** [review]
Rest LGTM. Thanks for the PR!

**Review by hzxa21 — COMMENTED** [review]
Generally LGTM after fixing all the comments.

File: src/storage/src/hummock/compactor/shared_buffer_compact.rs
Comment (StrikeW) [reply src/storage/src/hummock/compactor/shared_buffer_compact.rs]: Fixed. Add a trait `SharedBufferBatchIterCurrent` for the generic type `I` of `OrderedMergeIteratorInner`

File: src/storage/src/hummock/event_handler/hummock_event_handler.rs
Comment (StrikeW) [reply src/storage/src/hummock/event_handler/hummock_event_handler.rs:456]: If we cannot find the corresponding instance when we handle `ImmMerged`, the merged imm is useless. It is safe to drop it silently.

**Review by wenym1 — APPROVED** [review]

**Review by soundOfDestiny — COMMENTED** [review]
delete range isn't handled correctly.

**Review by soundOfDestiny — COMMENTED** [review]
delete range

**Pull request merged**
